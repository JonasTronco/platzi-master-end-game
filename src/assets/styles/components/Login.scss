@import '../main.scss';

.login {
  @include colCenCen;
  @include card;
  width: 320px;
  padding: $space_x5 $space_x6;

  &__logo {
    @include flexCenCen;
    width: 150px;
    height: 150px;
    margin: $space_x6;
    padding: 0;
    border-radius: 50%;
    background-color: $primary_color;

    &__icon {
      width: 90px;
      height: 90px;
      fill: $white_color;
    }
  }

  &__form {
    @include colCenCen;
    margin-bottom: $space_x6;

    .input {
      margin-bottom: $space_x3;
    }

    .input__label {
      @include colCenCen;
      width: 100%;
      text-align: center;
    }

    .btn {
      margin-top: $space_x3;
    }
  }
}

@media screen and (max-width: 425px) {
  .login {
    width: 100%;
    height: 100vh;
    border-radius: 0;
  }
}
