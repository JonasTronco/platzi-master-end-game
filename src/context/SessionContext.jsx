import { useState, useEffect, createContext, useMemo, useContext } from 'react';
import Router from 'next/router';

const SessionContext = createContext();
export function SessionProvider(props) {
  const [session, setSession] = useState(null);

  useEffect(() => {
    try {
      setSession(JSON.parse(localStorage.getItem('session')));
    } catch (error) {
      Router.push('/');
    }
  }, []);

  const value = useMemo(
    () => ({
      session,
    }),
    [session],
  );

  return <SessionContext.Provider value={value} {...props} />;
}

export function useSession() {
  const context = useContext(SessionContext);
  if (!context) {
    throw new Error('No Session Context Provider');
  }
  return context;
}
