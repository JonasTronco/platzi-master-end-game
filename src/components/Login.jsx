import { useState } from 'react';
import Router from 'next/router';
import axios from 'axios';
import { Formik, Form, Field, ErrorMessage } from 'formik';
import '../assets/styles/components/Login.scss';
import Logo from '../assets/icons/logo.svg';

function Login() {
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(null);
  const handleLoginSubmit = async (values) => {
    setLoading(true);
    const URL = `${process.env.NEXT_PUBLIC_API_URL}/auth/sign-in`;
    try {
      const response = await axios.post(
        URL,
        {
          apiKeyToken: process.env.NEXT_PUBLIC_API_KEY,
        },
        {
          auth: {
            username: values.username,
            password: values.password,
          },
        },
      );
      document.cookie = `session=${JSON.stringify(response.data)};max-age=${
        60 * 15
      }`;
      Router.push('/users');
    } catch (err) {
      setError(err);
    }
    setLoading(false);
  };

  if (error) {
    return <h1>Error</h1>;
  }

  return (
    <div className="login">
      <figure className="login__logo">
        {loading ? (
          <div className="loader loader--white" />
        ) : (
          <Logo className="login__logo__icon" />
        )}
      </figure>
      <Formik
        initialValues={{ username: '', password: '' }}
        validate={(values) => {
          const errors = {};
          if (!values.username) {
            errors.username = 'Required';
          }
          if (!values.password) {
            errors.password = 'Required';
          }
          return errors;
        }}
        onSubmit={async (values, { setSubmitting }) => {
          await handleLoginSubmit(values);
          setSubmitting(false);
        }}
      >
        {({ isSubmitting }) => (
          <Form className="login__form lg">
            <div className="input lg">
              <label className="input__label">
                Username / email
                <Field
                  className="input__field lg"
                  type="username"
                  name="username"
                  disabled={isSubmitting}
                />
              </label>
              <ErrorMessage
                name="username"
                component="div"
                className="input__error"
              />
            </div>
            <div className="input lg">
              <label className="input__label">
                Password
                <Field
                  className="input__field lg"
                  type="password"
                  name="password"
                  disabled={isSubmitting}
                />
              </label>
              <ErrorMessage
                name="password"
                component="div"
                className="input__error"
              />
            </div>
            <button className="btn lg" type="submit" disabled={isSubmitting}>
              Login
            </button>
          </Form>
        )}
      </Formik>
    </div>
  );
}

export default Login;
