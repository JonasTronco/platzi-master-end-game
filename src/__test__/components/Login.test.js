import Login from '../../components/Login';
import Router from 'next/router';
import { Formik, Form, Field, ErrorMessage } from 'formik';
import { create } from 'react-test-renderer';
import {
  render,
  cleanup,
  fireEvent,
  screen,
  getByTestId,
  act,
  RenderResult,
  wait
} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { SessionProvider } from '../../context/SessionContext';

describe('Login Component', () => {
  afterEach(cleanup);
  const mockedRouter = { push: () => {} };
  Router.router = mockedRouter;
  const obj = {
    session: {
      token:
        'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiI1ZWNkYTc2NjYxYjE3ZTUwYzg3NzIxYWIiLCJ1c2VybmFtZSI6Im1heW5lLnNuYXNlbC40MTcxIiwic2NvcGVzIjpbInNpZ25pbjphdXRoIiwic2lnbnVwOmF1dGgiLCJyZWFkOm1vdmllcyIsInJlYWQ6dXNlci1tb3ZpZXMiLCJjcmVhdGU6dXNlci1tb3ZpZXMiLCJkZWxldGU6dXNlci1tb3ZpZXMiXSwiaWF0IjoxNTkwNzI1MTIwLCJleHAiOjE1OTExNTcxMjB9.UjjAPEHWbqyPEwX6qcs2UhLvQ6rAeQIL-ME1LDnWBRo',
      user: {
        id: '5ecda76661b17e50c87721ab',
        username: 'mayne.snasel.4171',
        typeOfUser: 'Patient',
        isActive: true,
        imageURL: 'http://dummyimage.com/200x200.png/ff4444/ffffff',
        firstName: 'Mayne',
        lastName: 'Snasel',
        defaultPath: '/tests'
      }
    }
  };
  // before(function () {
  //   const button = button.querySelector('button');
  //   button.addEventListener(
  //     'click',
  //     () => {
  //       button;
  //     },
  //     true
  //   );
  // });

  test('Should render Login component', () => {
    const component = create(
      <SessionProvider value={obj}>
        <Login />
      </SessionProvider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('withFormik true input username renders fine', () => {
    const rendered = render(
      <SessionProvider value={obj}>
        <Login />
      </SessionProvider>
    );
    expect(rendered.getByTestId('username')).toBeInTheDocument();
    expect(rendered.container.getElementsByTagName('form')[0]).toBeDefined();
  });
  it('withFormik true button renders fine', () => {
    const rendered = render(
      <SessionProvider value={obj}>
        <Login />
      </SessionProvider>
    );
    expect(rendered.getByTestId('button-sign')).toBeInTheDocument();
    expect(rendered.container.getElementsByTagName('form')[0]).toBeDefined();
  });
});

// test('Should rendered of Login component', () => {
//   const login = render(<Login />);

//   expect(login.getByTestId('login-form')).toBeInTheDocument();
// });
// test('Testeando Login', () => {
//   expect(button).toHaveBeenCalledTimes(1);
// });

//   test('Hola', () => {
//     const handleLoginSubmit = jest.fn();
//     const component = render(<Login />);
//     act(() => {
//       fireEvent.click(component.getByTestId('button-sign'));
//     });
//     expect(handleLoginSubmit).toHaveBeenCalled();
//   });
// });

// describe('Test 2 of login', () => {
//   beforeEach(() => {
//     const handleLoginSubmit = jest.fn();
//     act(() => {
//       fireEvent.click(component.getByTestId('button-sign'));
//     });
//   });
// test('Hola', ()=>{
//       const component = render(<Login />);
//       expect(handleLoginSubmit).toHaveBeenCalled();

// })};
describe('Login test Component', () => {
  const onSubmit = jest.fn();

  beforeEach(() => {
    const onSubmit = jest.fn();
    act(() => {
      const component = render(<Login />);
      fireEvent.click(component.getByTestId('button-sign'));
    });
  });
  test('Should render login test component', () => {
    expect(onSubmit).toHaveBeenCalledTimes(1);
  });
});
