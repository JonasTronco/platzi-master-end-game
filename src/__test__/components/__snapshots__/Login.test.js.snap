// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component Should render Login component 1`] = `
<div
  className="login"
  data-testid="login-form"
>
  <figure
    className="login__logo"
  >
    <div
      className="login__logo__icon"
    />
  </figure>
  <form
    action="#"
    className="login__form lg"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="input lg"
    >
      <label
        className="input__label"
      >
        Username / email
        <input
          className="input__field lg"
          data-testid="username"
          disabled={false}
          name="username"
          onBlur={[Function]}
          onChange={[Function]}
          type="username"
          value=""
        />
      </label>
    </div>
    <div
      className="input lg"
    >
      <label
        className="input__label"
      >
        Password
        <input
          className="input__field lg"
          disabled={false}
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          type="password"
          value=""
        />
      </label>
    </div>
    <button
      className="btn lg"
      data-testid="button-sign"
      disabled={false}
      type="submit"
    >
      Login
    </button>
  </form>
</div>
`;
